### PR TITLE
162-param trained submission (1L, d=3, 3h, ff=6, fixed mask hybrid)

### PR DESCRIPTION
**Pull request by fblissjr**
## Summary

- **Category**: Trained (hybrid -- see note below)
- **Params**: 162
- **Accuracy**: 100% (10010/10010 on `verify.py --seed 2025`)
- **Architecture**: 1L decoder, d=3, 3 heads, ff=6
- **Built with**: Claude Code (Opus 4.6 with Sonnet 4.6 verifier subagents) + Gemini 3 Deep Think + my magic hands orchestrating like a conductor who is making it up as they go

## Approach: Fixed-Mask Hybrid

The attention mask is hand-designed (0 learnable parameters) -- it encodes WHERE to attend based on our hand-coded solution's 3-head routing:
- Head 0: digit pair extraction (A_k, B_k)
- Head 1: carry prefix sum with ALiBi slopes `-(k-j)*log(10)`
- Head 2: self-attend for residual management

The 162 trainable weights (embedding, Q/K/V/O projections, MLP, LM head) are learned via AdamW. The mask provides the routing structure; training learns WHAT to compute.

**Note on category**: The training algorithm (AdamW + CE loss) is generic, but the fixed attention mask is problem-specific. We're transparent about this -- happy to discuss whether this fits "Trained" or deserves a "Hybrid" designation. The mask is analogous to a fixed positional encoding (also not counted as params per the rules), but it encodes more structural information than typical PE.

## Training Details

- **Phase 1**: LR 0.01, 50K steps, cosine decay, batch 512, AdamW (wd=0.01)
- **Phase 2**: Load best checkpoint from Phase 1, LR 0.001, 5K steps
- **Data**: Random pairs generated on the fly (no fixed dataset)
- **Warm init**: Embedding initialized to `[digit_value, 0, 0]`, V-proj to `[[+1,0,0],[+1,0,0],[-1,0,0]]`, Q/K near zero

## Verification Output

```
Model: 162-Parameter Trained Adder
Author: fblissjr
Parameters (unique): 162
Architecture: 1L d=3 3h ff=6 reversed-digits

Results: 10010/10010 correct (100.00%)
Time: 19.6s (512 additions/sec)
Status: QUALIFIED (threshold: 99%)
```

## Link

- [submission_trained.py](https://github.com/fblissjr/AdderBoard/blob/main/submission_trained.py) (self-contained, weights inline)
- [train_adder.py](https://github.com/fblissjr/AdderBoard/blob/main/train_adder.py) (training script)
- [train_continue.py](https://github.com/fblissjr/AdderBoard/blob/main/train_continue.py) (fine-tuning script)
- [AGENTS.md](https://github.com/fblissjr/AdderBoard/blob/main/AGENTS.md) (full methodology and experiment logs)
- [Checkpoints](https://github.com/fblissjr/AdderBoard/tree/main/checkpoints) (all training artifacts)

### CODE REVIEW

**Review by anadim — COMMENTED** [review]
**Audit complete — cannot accept as "Trained"**

Verification results:
- **Accuracy**: 100.00% (10,010/10,010) — the model works perfectly
- **Parameter count**: 162 trainable verified

**However, the fixed attention mask is problematic for the Trained category.**

The mask (3,675 non-trainable constants) hardcodes the addition algorithm:
- Head 0 manually pairs digit A_k with B_k
- Head 1 implements carry prefix sum with ALiBi slopes -(k-j)*log(10)
- Head 2 handles residual cancellation

This carries ~1.76x more task-specific information than the 162 trained weights. The mask is derived directly from your own 33-param analytical solution. The rules state: "if your generation loop contains addition-specific logic — manually pairing digits, threading carry state, indexing into specific positions — then the Python code is solving the problem, not the model." The mask does exactly this, just relocated into a buffer.

The swap test also fails: you cannot use this architecture with different weights for a non-addition task — the mask constrains it to only addition.

**Your 33-param hand-coded submission (PR #54) is accepted and on the board.** That's the genuinely impressive result.

For this submission, I'd suggest one of:
1. Resubmit with a standard causal mask (no task-specific routing) — if training still works, it's legitimately trained
2. We could discuss creating a "Hybrid" category if there's community interest

Great transparency in flagging the hybrid nature yourself — really appreciate that.

**Review by anadim — COMMENTED** [review]
Thanks for the submission and for being transparent about the hybrid nature!

After thorough code review, we're not accepting this for the trained leaderboard. Here's why:

**The fixed mask is the algorithm, not a positional encoding.**

The `build_fixed_mask()` function encodes:
- **Head 0**: digit-pair extraction — routes each output digit to exactly its corresponding input digits A_k and B_k
- **Head 1**: carry prefix sum via `-(k-j)*log(10)` slopes — this is the base-10 carry propagation algorithm
- **Head 2**: self-attention only

This is 3,675 structured values that encode the full addition algorithm as attention routing. Standard positional encodings (RoPE, ALiBi, sinusoidal) are task-agnostic — they work for any sequence task. This mask only works for 10-digit addition with this specific input format.

The key evidence: without the mask, training peaks at ~35%. The mask is not an optimization — it *is* the solution. The 162 learned parameters only learn how to combine values that the mask already routes correctly.

Per Rule 4: "The model does the work, not the code." The `build_fixed_mask()` function is code doing the algorithmic work — it comes directly from your 33p hand-coded solution's routing pattern.

If we accepted this framing, the logical endpoint would be encoding any hand-coded solution as a fixed buffer and claiming 0 trained parameters.

Your 33-param hand-coded submission remains on the leaderboard and is a great entry! This hybrid approach is genuinely interesting from a research perspective (it cleanly separates routing from computation), but it doesn't fit the trained category rules as written.

**Comment by anadim** [comment]
Closing — see review comment above for details. Your 33p hand-coded entry remains on the board!
